File: native/android/app/src/main/java/com/colmhewson/crewbox/ScannerPlugin.java
package com.colmhewson.crewbox;

import android.Manifest;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.provider.Settings;

import androidx.activity.result.ActivityResult;

import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

/**
 * The join poster's QR, read with the camera, for the join screen
 * (web/src/components/Join.tsx), as CrewboxScanner.
 *
 * {@link ScannerActivity} does the reading, on the phone. What comes back is
 * the code's text and no more, and the page decides what that is, because a
 * QR is anybody's to print.
 *
 * {@code scan} resolves with what happened, for anything a crew member can
 * do: {@code scanned} with the text, {@code cancelled} when they back out,
 * {@code denied} when the camera isn't allowed, and {@code unavailable} when
 * the phone has no camera to use. It rejects only when the camera fails,
 * which the page words as the camera not starting.
 *
 * The camera is asked for here, the first time somebody scans. Once they
 * have said no twice Android stops asking and answers no at once, and the
 * page offers {@code openSettings}, the app's own page in Settings.
 *
 * Capacitor's own {@code checkPermissions} answers for the same "camera",
 * which "Take a photo" in the attach menu needs too: the message box asks it
 * when a photo comes back with nothing (web/src/components/Composer.tsx).
 */
@CapacitorPlugin(
    name = "CrewboxScanner",
    permissions = {@Permission(alias = "camera", strings = {Manifest.permission.CAMERA})})
public class ScannerPlugin extends Plugin {

  @PluginMethod
  public void scan(PluginCall call) {
    if (!getContext().getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)) {
      answer(call, "unavailable");
      return;
    }
    if (getPermissionState("camera") == PermissionState.GRANTED) {
      open(call);
      return;
    }
    requestPermissionForAlias("camera", call, "onCameraAnswer");
  }

  @PermissionCallback
  private void onCameraAnswer(PluginCall call) {
    if (getPermissionState("camera") == PermissionState.GRANTED) {
      open(call);
    } else {
      answer(call, "denied");
    }
  }

  private void open(PluginCall call) {
    startActivityForResult(call, new Intent(getContext(), ScannerActivity.class), "onScanned");
  }

  @ActivityCallback
  private void onScanned(PluginCall call, ActivityResult result) {
    if (call == null) return;
    getBridge().releaseCall(call);
    Intent data = result.getData();
    String text = data == null ? null : data.getStringExtra(ScannerActivity.EXTRA_TEXT);
    switch (result.getResultCode()) {
      case Activity.RESULT_OK:
        if (text == null) {
          answer(call, "cancelled");
          return;
        }
        JSObject scanned = new JSObject();
        scanned.put("result", "scanned");
        scanned.put("text", text);
        call.resolve(scanned);
        return;
      case ScannerActivity.RESULT_DENIED:
        answer(call, "denied");
        return;
      case ScannerActivity.RESULT_UNAVAILABLE:
        answer(call, "unavailable");
        return;
      case ScannerActivity.RESULT_FAILED:
        call.reject("The camera didn't start");
        return;
      default:
        answer(call, "cancelled");
    }
  }

  /** The app's own page in Settings, where the camera is allowed again. */
  @PluginMethod
  public void openSettings(PluginCall call) {
    Intent settings = new Intent(
        Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
        Uri.fromParts("package", getContext().getPackageName(), null));
    settings.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
    try {
      getContext().startActivity(settings);
      call.resolve();
    } catch (ActivityNotFoundException e) {
      call.reject("This phone has no settings page for the app");
    }
  }

  private static void answer(PluginCall call, String outcome) {
    JSObject result = new JSObject();
    result.put("result", outcome);
    call.resolve(result);
  }
}

File: native/ios/App/App/ScannerPlugin.swift
import AVFoundation
import Capacitor
import UIKit
import Vision
import VisionKit

/// The join poster's QR, read with the camera, for the join screen
/// (web/src/components/Join.tsx), as `window.Capacitor.Plugins.CrewboxScanner`.
///
/// VisionKit's DataScannerViewController does the reading. It needs iOS 16 and
/// an A12 chip or newer, which is every iPhone that runs iOS 17, this app's
/// minimum. It reads on the phone and sends nothing anywhere. What comes back
/// is the code's text and no more, and the page decides what that is, because
/// a QR is anybody's to print.
///
/// `scan` resolves with what happened, for anything a crew member can do:
/// `scanned` with the text, `cancelled` when they back out, `denied` when the
/// camera is off for Crewbox, and `unavailable` when this iPhone can't scan at
/// all, which includes a camera restricted by Screen Time or a profile: the
/// app's own page in Settings can't change that. It rejects only when the scan
/// can't be shown, which the page words as the camera not starting.
@objc(ScannerPlugin)
public class ScannerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ScannerPlugin"
    public let jsName = "CrewboxScanner"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "scan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openSettings", returnType: CAPPluginReturnPromise),
    ]

    /// The scan on screen. The scanner holds its delegate weakly, so this is
    /// what keeps the session alive until it finishes.
    private var session: ScanSession?

    @objc func scan(_ call: CAPPluginCall) {
        Task { @MainActor in self.start(call) }
    }

    /// The app's own page in Settings, where Camera is switched back on.
    @objc func openSettings(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString) else {
                call.reject("This iPhone has no Settings page for the app")
                return
            }
            UIApplication.shared.open(url) { _ in call.resolve() }
        }
    }

    @MainActor private func start(_ call: CAPPluginCall) {
        guard session == nil else {
            call.reject("A scan is already on screen")
            return
        }
        guard DataScannerViewController.isSupported else {
            call.resolve(["result": "unavailable"])
            return
        }
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            show(call)
        case .notDetermined:
            // Asked here rather than left to the scanner, so that the answer
            // comes back to this call whichever way it goes.
            AVCaptureDevice.requestAccess(for: .video) { granted in
                Task { @MainActor in
                    if granted {
                        self.show(call)
                    } else {
                        call.resolve(["result": "denied"])
                    }
                }
            }
        case .restricted:
            call.resolve(["result": "unavailable"])
        default:
            call.resolve(["result": "denied"])
        }
    }

    @MainActor private func show(_ call: CAPPluginCall) {
        // Available: supported, with the camera allowed and not restricted.
        guard DataScannerViewController.isAvailable else {
            call.resolve(["result": "unavailable"])
            return
        }
        guard let host = bridge?.viewController, host.presentedViewController == nil else {
            call.reject("Something else is on screen")
            return
        }
        let session = ScanSession { [weak self] outcome in
            self?.session = nil
            switch outcome {
            case let .scanned(text):
                call.resolve(["result": "scanned", "text": text])
            case .cancelled:
                call.resolve(["result": "cancelled"])
            case .denied:
                call.resolve(["result": "denied"])
            case .unavailable:
                call.resolve(["result": "unavailable"])
            }
        }
        self.session = session
        // Scanning starts once the screen is up, as Apple's own example does.
        // UIKit calls this back on the main thread, which is where the scan
        // lives, whether or not this SDK's signature says so.
        host.present(session.screen, animated: true) {
            MainActor.assumeIsolated { session.begin() }
        }
    }
}

/// What one scan came to.
private enum ScanOutcome {
    case scanned(String)
    case cancelled
    case denied
    case unavailable
}

/// One scan: the camera, full screen with Cancel, until it reads a QR with
/// text in it or is backed out of.
@MainActor
private final class ScanSession: NSObject, DataScannerViewControllerDelegate {
    let screen: UINavigationController
    private let scanner: DataScannerViewController
    private var done: ((ScanOutcome) -> Void)?
    private var torch: UIBarButtonItem?
    private var torchOn = false

    init(done: @escaping (ScanOutcome) -> Void) {
        // QR only: the poster prints nothing else, and any other kind of code
        // would be something else on the same wall.
        scanner = DataScannerViewController(
            recognizedDataTypes: [.barcode(symbologies: [.qr])],
            qualityLevel: .balanced,
            recognizesMultipleItems: false,
            isHighFrameRateTrackingEnabled: false,
            isPinchToZoomEnabled: true,
            isGuidanceEnabled: true,
            isHighlightingEnabled: true
        )
        screen = UINavigationController(rootViewController: scanner)
        self.done = done
        super.init()
        scanner.delegate = self
        scanner.title = "Scan the join poster"
        scanner.navigationItem.leftBarButtonItem = UIBarButtonItem(
            systemItem: .cancel,
            primaryAction: UIAction { [weak self] _ in self?.finish(.cancelled) }
        )
        // A poster in a dark tent needs light on it.
        if AVCaptureDevice.default(for: .video)?.hasTorch == true {
            let torch = UIBarButtonItem(
                image: UIImage(systemName: "flashlight.off.fill"),
                primaryAction: UIAction { [weak self] _ in self?.toggleTorch() }
            )
            torch.accessibilityLabel = "Torch"
            torch.accessibilityValue = "Off"
            scanner.navigationItem.rightBarButtonItem = torch
            self.torch = torch
        }
        // A bar with its own background: Cancel has to read over whatever the
        // camera is pointed at, in daylight.
        let bar = UINavigationBarAppearance()
        bar.configureWithDefaultBackground()
        screen.navigationBar.standardAppearance = bar
        screen.navigationBar.scrollEdgeAppearance = bar
        // Full screen: a sheet can be swiped away past Cancel, and then the
        // page would never hear how the scan ended.
        screen.modalPresentationStyle = .fullScreen
    }

    func begin() {
        do {
            try scanner.startScanning()
        } catch let error as DataScannerViewController.ScanningUnavailable {
            finish(Self.outcome(of: error))
        } catch {
            finish(.unavailable)
        }
    }

    func dataScanner(
        _ dataScanner: DataScannerViewController,
        didAdd addedItems: [RecognizedItem],
        allItems: [RecognizedItem]
    ) {
        for item in addedItems {
            // A QR can carry bytes that aren't text, which no poster prints:
            // the camera stays up for the one that is.
            if case let .barcode(code) = item, let text = code.payloadStringValue, !text.isEmpty {
                finish(.scanned(text))
                return
            }
        }
    }

    func dataScanner(
        _ dataScanner: DataScannerViewController,
        becameUnavailableWithError error: DataScannerViewController.ScanningUnavailable
    ) {
        finish(Self.outcome(of: error))
    }

    /// Restricted is Apple's word for the camera not being allowed, whether
    /// by the crew member or by a profile.
    private static func outcome(of error: DataScannerViewController.ScanningUnavailable)
        -> ScanOutcome
    {
        if case .cameraRestricted = error { return .denied }
        return .unavailable
    }

    private func toggleTorch() {
        setTorch(!torchOn)
    }

    private func setTorch(_ on: Bool) {
        guard let device = AVCaptureDevice.default(for: .video), device.hasTorch,
            device.isTorchAvailable, (try? device.lockForConfiguration()) != nil
        else { return }
        device.torchMode = on ? .on : .off
        device.unlockForConfiguration()
        torchOn = on
        torch?.image = UIImage(systemName: on ? "flashlight.on.fill" : "flashlight.off.fill")
        torch?.accessibilityValue = on ? "On" : "Off"
    }

    private func finish(_ outcome: ScanOutcome) {
        guard let done else { return }
        self.done = nil
        if torchOn { setTorch(false) }
        scanner.stopScanning()
        if case .scanned = outcome {
            UINotificationFeedbackGenerator().notificationOccurred(.success)
        }
        screen.presentingViewController?.dismiss(animated: true)
        done(outcome)
    }
}
